src/core/Akka/Actor/Futures.cs

```diff
@@ -108,7 +111,17 @@ public static Task<T> Ask<T>(this ICanTell self, object message, CancellationTok
             if (provider == null)
                 throw new ArgumentException("Unable to resolve the target Provider", nameof(self));
 
-            return (T) await Ask(self, message, provider, timeout, cancellationToken);
+            var result = await Ask(self, message, provider, timeout, cancellationToken);
+            
+            // if failure was delivered and not expected, rethrow it
+            if (result is Status.Failure f && typeof(T) != FailureType)
+            {
+                // if failure was result of cancelled task, exception will be null
+                var cause = f.Cause ?? new TaskCanceledException();
+                ExceptionDispatchInfo.Capture(cause).Throw();
+                return default(T);
+            }
+            else return (T) result;
```

Comment:
I am wondering if it's a good idea in general. If something failed somewhere - would not it be better not to reply at all? Basically we are speaking of unhandled exception... Maybe it's better let it die there?

ps. will it work across remote ?

Reply:
@maxcherednik it will work across remote connections as long as serializer is able to correctly serialize/deserialize exceptions (I'm not sure what's the status of it on .NET core). Right now if something has failed and `Status.Failure` is returned, it's the worst scenario of all: the InvalidCastExcepton is throw (as we try to cast Status.Failure to T), while the original reason is lost. I've already tried so fix that back in #3227 , but I've thought that your work over `Ask` upgrade will cover that.

Reply:
I see. No, did not cover that case. InvalidCastException is ugly for sure.

But do we need to bring some deep exception, which arrived across async message boundary?  As an option - we don't catch unhandled exception at all and asking part will just see that nothing was sent back for whatever reason(network, slowness, db issue). Anything which is not public API of that remote service should not be sent back. IMHO.

Reply:
I think, this case (ignore sending failures back to the user) should be handled on the sender side as opt-in. I.e. if `PipeTo(Sender, failure: ex => null)`, then don't send any response back.

Reply:
@Horusiath @Aaronontheweb what I am trying to say, Status.Failure is not any actor's default API in case of error, right? And it should not be a special case in Ask method. I remember, I did propagate a general Exception somewhere - that time I cringe and thought, I probably should not catch it at all, it should fail and die.